### PR TITLE
Display user friendly messages for Json Schema validation failures

### PR DESCRIPTION
**Pull request by rishi-aga**
related to PR: https://github.com/yahoo/elide/pull/1662

## Description
Remove singletons. Add Javadoc.
Add custom messages for queryPlanResolver & filterTemplate.
Enforce no additional properties for dimensions and time dimensions.

## How Has This Been Tested?
Existing tests pass.
Additional tests included.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DraftV4LibraryWithElideFormatAttr.java [hunk 1]
```diff
@@ -5,46 +5,41 @@
...
 import com.github.fge.jsonschema.library.DraftV4Library;
 import com.github.fge.jsonschema.library.Library;
 import com.github.fge.jsonschema.library.LibraryBuilder;
 
-import lombok.NoArgsConstructor;
+import lombok.Getter;
 
-@NoArgsConstructor(access = lombok.AccessLevel.PRIVATE)
 public class DraftV4LibraryWithElideFormatAttr {
-    private static Library LIBRARY = null;
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DraftV4LibraryWithElideFormatAttr.java:24]: Let's add javadoc for all of these new classes.

File: elide-model-config/src/main/resources/elideTableSchema.json [hunk 2]
```diff
@@ -191,22 +191,7 @@
...
-                        },
-                        "tableSource": {
-                            "not": {
-                                "required": [
-                                    "values"
-                                ]
-                            }
-                        }
-                    }
+                    "eitherTableSourceOrValues": {}
```
Comment (aklish) [inline elide-model-config/src/main/resources/elideTableSchema.json:194]: Why did we do this swap?  Was the error hard to follow?

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideRSQLFilterFormatAttr.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.github.fge.jackson.NodeType;
+import com.github.fge.jsonschema.core.exceptions.ProcessingException;
+import com.github.fge.jsonschema.core.report.ProcessingReport;
+import com.github.fge.jsonschema.format.AbstractFormatAttribute;
+import com.github.fge.jsonschema.processors.data.FullData;
+import com.github.fge.msgsimple.bundle.MessageBundle;
+
+public class ElideRSQLFilterFormatAttr extends AbstractFormatAttribute {
+    private static final String RSQL_FILTER_FORMAT_REGEX = "^[A-Za-z][0-9A-Za-z_]*(==|!=|>=|>|<|<=|=[a-z]+=)(.*)$";
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideRSQLFilterFormatAttr.java:16]: I wonder if we can do something stronger here.  I wonder if we can parse the expression with FIQL/RSQL parser.  A regex is not expressive enough to capture all the legal values.

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideRSQLFilterFormatAttr.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    public static final String FORMAT_NAME = "elideRSQLFilter";
+    public static final String FORMAT_KEY = "elideRSQLFilter.error.format";
+    public static final String FORMAT_MSG = "filterTemplate [%s] is not allowed. "
+                    + "RSQL filter Template must follow the format 'XoperatorY;XoperatorY;XoperatorY'. "
+                    + "Here `X` must start with an alphabet and can include alaphabets, numbers and '_' only. "
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideRSQLFilterFormatAttr.java:22]: This comment doesn't capture the breadth of what the user can define.  The semi-colon can also be a comma.  Each operator has its own rules. 

I think a pointer to the RSQL documentation might be better than trying to redefine here.

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideClassNameFormatAttr.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.github.fge.jackson.NodeType;
+import com.github.fge.jsonschema.core.exceptions.ProcessingException;
+import com.github.fge.jsonschema.core.report.ProcessingReport;
+import com.github.fge.jsonschema.format.AbstractFormatAttribute;
+import com.github.fge.jsonschema.processors.data.FullData;
+import com.github.fge.msgsimple.bundle.MessageBundle;
+
+public class ElideClassNameFormatAttr extends AbstractFormatAttribute {
+    private static final String CLASS_NAME_FORMAT_REGEX = "^([a-zA-Z]+[a-zA-Z0-9_]*\\.)+class$";
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideClassNameFormatAttr.java:16]: This might have been too simplistic.  Check out:
https://stackoverflow.com/questions/5205339/regular-expression-matching-fully-qualified-class-names

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideClassNameFormatAttr.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    public static final String FORMAT_NAME = "elideClassName";
+    public static final String FORMAT_KEY = "elideClassName.error.format";
+    public static final String FORMAT_MSG = "Class Name [%s] is not allowed. "
+                    + "Class name must follow the format 'X.X.X.X.class'. "
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideClassNameFormatAttr.java:21]: Something simple here might be better:  "[%s] is not a valid Java class name."

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesValidator.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import com.google.common.collect.Sets;
+
+import java.util.Set;
+
+public class EitherTableSourceOrValuesValidator extends AbstractKeywordValidator {
+
+    public static final String KEYWORD = "eitherTableSourceOrValues";
+    public static final String ERROR_KEY = "eitherTableSourceOrValues.error";
+    public static final String ERROR_MSG =
+                    "Either tableSource or values should be defined for a dimension, Both are not allowed.";
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesValidator.java:24]: Neither of them is also an option.

I would reword: "tableSource and values cannot both be defined for a dimension.  Choose one or none."

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesValidator.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        JsonNode instance = data.getInstance().getNode();
+        Set<String> fields = Sets.newHashSet(instance.fieldNames());
+
+        if (fields.contains("values") && fields.contains("tableSource")) {
+            report.error(newMsg(data, bundle, ERROR_KEY));
+        }
+    }
+
+    @Override
+    public String toString() {
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesValidator.java:43]: Why do we define toString here but nowhere else?

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesKeyword.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    private class EitherTableSourceOrValuesSyntaxChecker extends AbstractSyntaxChecker {
+
+        public EitherTableSourceOrValuesSyntaxChecker() {
+            super(EitherTableSourceOrValuesValidator.KEYWORD, NodeType.OBJECT);
+        }
+
+        @Override
+        protected void checkValue(Collection<JsonPointer> pointers, MessageBundle bundle, ProcessingReport report,
+                        SchemaTree tree) throws ProcessingException {
+            // No Checks Required
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesKeyword.java:45]: Do we need to validate that we have an object node here?  Do we need to check for fields that are not tableSource or values?

**Review by aklish — CHANGES_REQUESTED** [review]

File: elide-model-config/src/main/resources/elideTableSchema.json (hunk 2 above)
Comment (rishi-aga) [reply elide-model-config/src/main/resources/elideTableSchema.json]: yes.

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DraftV4LibraryWithElideFormatAttr.java (hunk 1 above)
Comment (rishi-aga) [reply elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DraftV4LibraryWithElideFormatAttr.java:24]: added.

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideClassNameFormatAttr.java (hunk 5 above)
Comment (rishi-aga) [reply elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideClassNameFormatAttr.java]: updated.

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideClassNameFormatAttr.java (hunk 6 above)
Comment (rishi-aga) [reply elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideClassNameFormatAttr.java]: updated.

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesValidator.java (hunk 7 above)
Comment (rishi-aga) [reply elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesValidator.java]: updated.

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesValidator.java (hunk 8 above)
Comment (rishi-aga) [reply elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesValidator.java]: this one extends `AbstractKeywordValidator` which has abstract `toString` method. Others extends `AbstractFormatAttribute` which doesn't have `toString` method.

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesKeyword.java (hunk 9 above)
Comment (rishi-aga) [reply elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/EitherTableSourceOrValuesKeyword.java]: Constructor verifies the type of node.

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideRSQLFilterFormatAttr.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+ * <p>
+ * This specifier will check if a string instance is a valid RSQL filter.
+ * </p>
+ */
+public class ElideRSQLFilterFormatAttr extends AbstractFormatAttribute {
+
+    public static final String FORMAT_NAME = "elideRSQLFilter";
+    public static final String FORMAT_KEY = "elideRSQLFilter.error.format";
+    public static final String FORMAT_MSG = "Input value[%s] is not a valid RSQL filter expression. Please visit page "
+                    + "https://elide.io/pages/guide/v5/11-graphql.html and search 'RSQL operators' for samples.";
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideRSQLFilterFormatAttr.java:30]: You can provide a URL fragment (https://something.com#fragment) to link to a section in the documentation.

**Review by aklish — CHANGES_REQUESTED** [review]

File: elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideRSQLFilterFormatAttr.java (hunk 10 above)
Comment (rishi-aga) [reply elide-model-config/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/jsonformats/ElideRSQLFilterFormatAttr.java]: updated.

**Review by aklish — APPROVED** [review]
